pkg/resource/deploy/source_eval.go

```diff
@@ -782,6 +797,41 @@ func parseProviderRequest(
 	return providers.NewProviderRequest(&parsedVersion, pkg, url, pluginChecksums), nil
 }
 
+func (rm *resmon) RegisterProvider(ctx context.Context,
+	req *pulumirpc.RegisterProviderRequest,
+) (*pulumirpc.RegisterProviderResponse, error) {
+	logging.V(5).Infof("ResourceMonitor.RegisterProvider(%v)", req)
+
+	// First parse the request into a ProviderRequest
+	var version *semver.Version
+	if req.Version != "" {
+		v, err := semver.Parse(req.Version)
+		if err != nil {
+			return nil, fmt.Errorf("parse package version %s: %w", req.Version, err)
+		}
+		version = &v
+	}
+	pi := providers.NewProviderRequest(
+		version, tokens.Package(req.Name), req.PluginDownloadUrl, req.PluginChecksums)
+
+	rm.packageRefLock.Lock()
```

Comment:
I was wondering if it was worth using a readlock here and then upgrading to a writelock later.  But iiuc this only needs to be called once per component, not once per resonrce, right?  So this should not be called very often per pulumi program, and should be okay to just have this one lock, and having a read lock is a bit more tricky anyway.

Reply:
yeh once per process generally